packages/eslint-plugin/src/rules/no-unused-vars.ts

```diff
@@ -557,40 +560,71 @@ export default createRule<Options, MessageIds>({
     }
 
     return {
-      // declaration file handling
-      [ambientDeclarationSelector(AST_NODE_TYPES.Program, true)](
+      // top-level declaration file handling
+      [ambientDeclarationSelector(AST_NODE_TYPES.Program)](
         node: DeclarationSelectorNode,
       ): void {
         if (!isDefinitionFile(context.filename)) {
           return;
         }
+
+        const moduleDecl = nullThrows(
+          node.parent,
+          NullThrowsReasons.MissingParent,
+        ) as TSESTree.Program;
+
+        if (checkForOverridingExportStatements(moduleDecl)) {
+          return;
+        }
+
         markDeclarationChildAsUsed(node);
       },
 
       // children of a namespace that is a child of a declared namespace are auto-exported
       [ambientDeclarationSelector(
         'TSModuleDeclaration[declare = true] > TSModuleBlock TSModuleDeclaration > TSModuleBlock',
-        false,
       )](node: DeclarationSelectorNode): void {
+        const moduleDecl = nullThrows(
+          node.parent.parent,
+          NullThrowsReasons.MissingParent,
+        ) as TSESTree.TSModuleDeclaration;
+
+        if (checkForOverridingExportStatements(moduleDecl)) {
+          return;
+        }
+
         markDeclarationChildAsUsed(node);
       },
 
       // declared namespace handling
       [ambientDeclarationSelector(
         'TSModuleDeclaration[declare = true] > TSModuleBlock',
-        false,
       )](node: DeclarationSelectorNode): void {
         const moduleDecl = nullThrows(
           node.parent.parent,
           NullThrowsReasons.MissingParent,
         ) as TSESTree.TSModuleDeclaration;
 
-        // declared ambient modules with an `export =` statement will only export that one thing
-        // all other statements are not automatically exported in this case
-        if (
-          moduleDecl.id.type === AST_NODE_TYPES.Literal &&
-          checkModuleDeclForExportEquals(moduleDecl)
-        ) {
+        if (checkForOverridingExportStatements(moduleDecl)) {
+          return;
+        }
+
+        markDeclarationChildAsUsed(node);
+      },
+
+      // namespace handling in definition files
+      [ambientDeclarationSelector('TSModuleDeclaration > TSModuleBlock')](
+        node: DeclarationSelectorNode,
+      ): void {
+        if (!isDefinitionFile(context.filename)) {
+          return;
+        }
+        const moduleDecl = nullThrows(
+          node.parent.parent,
+          NullThrowsReasons.MissingParent,
+        ) as TSESTree.TSModuleDeclaration;
+
+        if (checkForOverridingExportStatements(moduleDecl)) {
           return;
         }
```

Comment:
This addition is for making not reported the following ([playground link](https://typescript-eslint.io/play/#ts=5.7.3&fileType=.d.ts&code=KYDwDg9gTgLgBAOwIYFtgGcxIMbDgMQgjgG8AoOOAeirimElmABM4AjAV3nQAsIOANswQByGBTjYICdPABmRAFxwAjACYAzABYA3GQC%2BZIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tidmUQAmtAG4BDaKgyRE0aB2iRwYAL4hlQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)):

```ts
export namespace Foo {
  // reported but shouldn't
  const foo: 1234;
}
```